core/base/combination.hpp

```diff
@@ -0,0 +1,171 @@
+
+public:
+    using value_type = ValueType;
+
+protected:
+    /**
+     * Creates an empty linear combination (0x0 operator).
+     *
+     * @param exec  Executor associated to the linear combination
+     */
+    explicit Combination(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Combination>(exec)
+    {}
+
+    /**
+     * Creates a linear combination of operators using the specified list of
+     * coefficients and operators.
+     *
+     * @tparam CoefficientIterator  a class representing iterators over the
+     *                              coefficients of the linear combination
+     * @tparam OperatorIterator  a class representing iterators over the
+     *                           operators of the linear combination
+     *
+     * @param coefficient_begin  iterator pointing to the first coefficient
+     * @param coefficient_end  iterator pointing behind the last coefficient
+     * @param operator_begin  iterator pointing to the first operator
+     * @param operator_end  iterator pointing behind the last operator
+     */
+    template <
+        typename CoefficientIterator, typename OperatorIterator,
+        typename = xstd::void_t<
+            typename std::iterator_traits<
+                CoefficientIterator>::iterator_category,
+            typename std::iterator_traits<OperatorIterator>::iterator_category>>
+    explicit Combination(CoefficientIterator coefficient_begin,
+                         CoefficientIterator coefficient_end,
+                         OperatorIterator operator_begin,
+                         OperatorIterator operator_end)
+        : EnableLinOp<Combination>(
+              first_or_throw(operator_begin, operator_end)->get_executor()),
+          coefficients_(coefficient_begin, coefficient_end),
+          operators_(operator_begin, operator_end)
+    {
+        for (const auto &c : coefficients_) {
+            ASSERT_EQUAL_DIMENSIONS(c, dim<2>(1, 1));
+        }
+        this->set_size(operators_[0]->get_size());
+        for (const auto &o : operators_) {
+            ASSERT_EQUAL_DIMENSIONS(o, this->get_size());
+        }
+    }
+
+    /**
+     * Creates a linear combination of operators using the specified list of
+     * coefficients and operators.
+     *
+     * @tparam Rest  types of trailing parameters
+     *
+     * @param coefficients  the first coefficient
+     * @param operators  the first operator
+     * @param rest  other coefficient and operators (interleaved)
+     */
+    template <typename... Rest>
+    explicit Combination(std::shared_ptr<const LinOp> coef,
+                         std::shared_ptr<const LinOp> oper, Rest &&... rest)
+        : Combination(std::forward<Rest>(rest)...)
+    {
+        ASSERT_EQUAL_DIMENSIONS(coef, dim<2>(1, 1));
+        ASSERT_EQUAL_DIMENSIONS(oper, this->get_size());
+        coefficients_.insert(begin(coefficients_), coef);
+        operators_.insert(begin(operators_), oper);
```

Comment:
I have one question here, what happens if there are only 3 LinOps given as arguments to the constructor, does it not compile? If it does compile (how?) then we need to make sure an error is thrown or something.

Reply:
It doesn't compile. When trying to resolve the "recursive" (quotes because there's no actual recursion here, a different constructor is called in every "recursive" step) constructor being called, the last step will fails, as it won't find a constructor that has only one `LinOp` as the parameter.